build.py

```diff
@@ -341,6 +344,91 @@ def step_end():
     step_end()
 
+if build_notebooks:
+    step_start("Running notebook samples")
+    # Find all notebooks in the samples directory. Skip the basic sample and the azure submission sample, since those won't run
+    # nicely in automation.
+    notebook_files = [os.path.join(dp, f) for dp, _, filenames in os.walk(samples_src) for f in filenames if f.endswith(".ipynb")
+                      and not (f.startswith("sample.") or f.startswith("azure_submission."))]
+    # Check if in a virtual environment
+    if (
+        os.environ.get("VIRTUAL_ENV") is None
+        and os.environ.get("CONDA_PREFIX") is None
+        and os.environ.get("CI") is None
+    ):
+        print("Not in a virtual python environment")
+
+        venv_dir = os.path.join(samples_src, ".venv")
+        # Create virtual environment under repo root
+        if not os.path.exists(venv_dir):
+            print(f"Creating a virtual environment under {venv_dir}")
+            venv.main([venv_dir])
+
+        # Check if .venv/bin/python exists, otherwise use .venv/Scripts/python.exe (Windows)
+        python_bin = os.path.join(venv_dir, "bin", "python")
+        if not os.path.exists(python_bin):
+            python_bin = os.path.join(venv_dir, "Scripts", "python.exe")
+        print(f"Using python from {python_bin}")
+    else:
+        # Already in a virtual environment, use current Python
+        python_bin = sys.executable
+
+    # copy the process env vars
+    pip_env: dict[str, str] = os.environ.copy()
+
+    # Install the qsharp package
+    pip_install_args = [
+        python_bin,
+        "-m",
+        "pip",
+        "install",
+        "-e",
+        pip_src,
+    ]
+    subprocess.run(pip_install_args, check=True, text=True, cwd=pip_src, env=pip_env)
+
+    # Install the widgets package
+    pip_install_args = [
+        python_bin,
+        "-m",
+        "pip",
+        "install",
+        "-e",
+        widgets_src,
+    ]
+    subprocess.run(pip_install_args, check=True, text=True, cwd=widgets_src, env=pip_env)
+
+    # Install other dependencies
+    pip_install_args = [
+        python_bin,
+        "-m",
+        "pip",
+        "install",
+        "ipykernel",
+        "nbconvert",
+        "pandas",
+    ]
+    subprocess.run(pip_install_args, check=True, text=True, cwd=root_dir, env=pip_env)
+
+    for notebook in notebook_files:
+        print(f"Running {notebook}")
+        # Run the notebook process, capturing stdout and only displaying it if there is an error
+        result = subprocess.run([python_bin,
+                        "-m",
+                        "nbconvert",
+                        "--to",
+                        "notebook",
+                        "--stdout",
+                        "--ExecutePreprocessor.timeout=60",
+                        "--execute",
+                        notebook],
+                        check=False, text=True, cwd=root_dir, env=pip_env, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
```

Comment:
Platform dependencies rear their ugly head again... it seems like one of the notebooks produces output that the Python infrastructure on Windows can't convert into a character string. I need to figure out how to adjust for that without needing to go back to spamming the logs with the very, very long output of the converted notebook.

Reply:
Ok, explicitly specifying encoding seems to have resolved the issue. Apparently Windows is the only platform that doesn't set that by default.